src/aiida_quantumespresso/workflows/pw/relax.py

```diff
@@ -192,74 +198,134 @@ def setup(self):
+        workchain = self.ctx.base_init_relax_workchain
+
+        if not workchain.is_finished_ok:
+            self.report(f'final scf PwBaseWorkChain failed with exit status {workchain.exit_status}')
+            return self.exit_codes.ERROR_SUB_PROCESS_FAILED_INIT_RELAX
+
+        self.ctx.current_structure = workchain.outputs.output_structure
+        self.ctx.current_number_of_bands = workchain.outputs.output_parameters.get_dict()['number_of_bands']
 
     def should_run_relax(self):
         """Return whether a relaxation workchain should be run.
 
-        This is the case as long as the volume change between two consecutive relaxation runs is larger than the volume
-        convergence threshold value and the maximum number of meta convergence iterations is not exceeded.
-        """
-        return not self.ctx.is_converged and self.ctx.iteration < self.inputs.max_meta_convergence_iterations.value
+        This is the case as long as either:
 
-    def should_run_final_scf(self):
-        """Return whether after successful relaxation a final scf calculation should be run.
+        1. There are still Pulay stresses in the system. When running a `vc-relax`, Quantum ESPRESSO will automatically
+           run a final SCF after the geometry optimization, where it regenerates the plane wave basis sets using the
+           final lattice of the final geometry. If the stresses are too large in this final SCF, this will means there
+           are still Pulay stresses and the `PwBaseWorkChain` will check for this and return exit code 501.
+        2. The k-points mesh was defined as a density using `kpoints_distance` for the `base_relax` input name space,
+           and the unit cell has changed in such a way that regenerating the k-points mesh using this density increases
+           the mesh compared to the previous run.
 
-        If the maximum number of meta convergence iterations has been exceeded and convergence has not been reached, the
-        structure cannot be considered to be relaxed and the final scf should not be run.
+        If one of these two conditions is met, another relaxation should be run as long as the maximum number of meta
+        convergence iterations is not exceeded.
         """
-        return self.ctx.is_converged and 'final_scf_inputs' in self.ctx
+        # If no work chain has been run, we should at least run one
+        if 'base_relax_workchains' not in self.ctx:
+            return True
+
+        # If meta convergence is switched off we are done
+        if not self.ctx.meta_convergence:
+            self.report('Meta-convergence disabled: workchain completed after a single iteration.')
+            return False
+
+        # Stop if the maximum number of meta iterations has been reached
+        if self.ctx.iteration == self.inputs.max_meta_convergence_iterations.value:
+            self.report('Maximum number of meta convergence iterations reached.')
+            return False
```

Comment:
I was looking for the logic where this would set an exit code to say the workchain didn't converge, but to my surprise it is not there, and also didn't use to be there. Think we'd better add one, right?